### DISCUSSION TITLE
Form submission all in the same route and isDirty

### DISCUSSION THREAD

**Discussion by caprica** [Code-Forge-Net/remix-hook-form#72] [Q&A]
I am having some trouble understanding how to deal with a dirty form after submission.

In a route I have:

- a Form with default values
- an input where I change the value, so the form state changes isDirty to true
- it submits to an action in the same route
- the action returns null
- the loaders execute
- the page updates with new default values for the form, I check them via the getValues form method

(Tthere are no redirects or route changes.)

BUT, the form still reports isDirty = true.

If I invoke reset on the form in response to a change in the isSubmitSuccessful form state, the inputs reset to the previously unchanged values, which is not what I want.

If I add a useEffect with a dependency on the default values, it DOES reset the form and the dirty status, but there is a momentary flash of it changing first to the old values, which is not nice from a UX perspective, and this seemed like a hack tbh.

I suppose I expected that the new default values would trigger the form to reset, since in principle the form values now match the values in the inputs. It seems like the form inputs are somehow not quite in sync with the new default values, even though they appear to be the same.

Anyone has any suggestions or a pointer to a similar example?

This is similar to, but not the same as, #60 and #62.

**Comment by caprica** [comment]
In theory, this is supposed to work at least with react-hook-form, by keeping the submitted values when `reset()` is later invoked:

```typescript
const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isSubmitSuccessful },
} = useRemixForm<Inputs>({ resolver, defaultValues: data.formData, resetOptions: { keepValues: true } })

useEffect(() => {
    if (isSubmitSuccessful) {
        reset()
    }
}, [isSubmitSuccessful, reset])
```

But that does not work for me with remix-hook-form, `reset()` will result in the *original* values being restored.

This was the only way I could get the behaviour I want, using `getValues()` in conjunction with `reset()`:

```typescript
const {
    register,
    handleSubmit,
    reset,
    getValues,
    formState: { errors, isDirty, isSubmitSuccessful },
} = useRemixForm<Inputs>({ resolver, defaultValues: data.formData })

useEffect(() => {
    if (isSubmitSuccessful) {
        reset(getValues())
    }
}, [isSubmitSuccessful, reset, getValues])
```

(ref: https://github.com/react-hook-form/react-hook-form/issues/3097#issuecomment-1299924673)

**Comment by caprica** [comment]
Added a sandbox here: https://codesandbox.io/p/devbox/silly-jepsen-gzp9mr

**Comment by theskillwithin** [comment]
this should be an issue?

**Comment by nboliver-ventureweb** [comment]
@caprica Thanks for sharing your solution. I was looking to do the same, however I found that `isSubmitSuccessful` isn't working - it always stays false. My form uses a `useSubmit` because the Button component in Polaris doesn't have a `form` attribute and the button needs to be located outside the form element. I ended up passing the success state as a prop via `navigation.state === 'idle' && actionData?.ok`. Would be great to better understand what parts of remix-hook-form don't work the same as react-hook-form.
